File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Columns/StoreListColumn.cs
using CitizenFX.Core;
using ScaleformUI.Elements;
using ScaleformUI.LobbyMenu;
using ScaleformUI.Menu;
using ScaleformUI.Menus;
using ScaleformUI.PauseMenu;
using ScaleformUI.PauseMenus.Elements.Items;
using System.Linq;

namespace ScaleformUI.PauseMenus.Elements.Columns
{
    public delegate void StoreItemSelected(StoreItem item, int index);
    public class StoreListColumn : PM_Column
    {
        public event IndexChanged OnIndexChanged;
        private List<PauseMenuItem> _unfilteredItems;
        private int _unfilteredSelection = 0;
        public event StoreItemSelected StoreItemActivated;
        public StoreListColumn(string label) : base(-1)
        {
            Label = label;
            VisibleItems = 4;
            type = (int)PLT_COLUMNS.STORE;
        }

        public void AddStoreItem(StoreItem item)
        {
            item.ParentColumn = this;
            Items.Add(item);
            if (visible && Items.Count <= VisibleItems)
            {
                var idx = Items.Count - 1;
                AddSlot(idx);
                item.Selected = idx == index;
            }
        }

        public override void SetDataSlot(int index)
        {
            this.SendItemToScaleform(index);
        }

        public override void UpdateSlot(int index)
        {
            if (index >= Items.Count) return;
            if (visible)
                SendItemToScaleform(index, true);
        }

        public override void AddSlot(int index)
        {
            if (index >= Items.Count) return;
            if (visible)
                SendItemToScaleform(index, false, false, true);
        }

        public override void Populate()
        {
            Main.PauseMenu._pause.CallFunction("SET_DATA_SLOT_EMPTY", (int)position);
            Main.PauseMenu._pause.CallFunction("SET_COLUMN_MAX_ITEMS", (int)position, VisibleItems);
            for (var i = 0; i < Items.Count; i++)
            {
                SetDataSlot(i);
            }
        }

        internal void SendItemToScaleform(int i, bool update = false, bool newItem = false, bool isSlot = false)
        {
            if (i >= Items.Count) return;
            StoreItem item = (StoreItem)Items[i];
            string str = "SET_DATA_SLOT";
            if (update)
                str = "UPDATE_SLOT";
            if (newItem)
                str = "ADD_SLOT";
            if (isSlot)
                str = "SET_SLOT_EMPTY";
            Main.PauseMenu._pause.CallFunction(str, (int)position, i, 0, 0, i, 0, item.Enabled, item.textureDictionary, item.textureName, item.Description);
        }

        public override void ShowColumn(bool show = true)
        {
            base.ShowColumn(show);
            InitColumnScroll(true, 1, ScrollType.UP_DOWN, ScrollArrowsPosition.RIGHT);
            SetColumnScroll(Index + 1, Items.Count, VisibleItems, CaptionLeft, Items.Count < VisibleItems);
            Main.PauseMenu._pause.CallFunction("SET_COLUMN_FOCUS", (int)position, Focused, false, false);
        }

        public void Clear()
        {
            if (visible)
                ClearColumn();
            Items.Clear();
        }

        public override void GoUp()
        {
            try
            {
                CurrentItem.Selected = false;
                index--;
                if (index < 0)
                    index = Items.Count - 1;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_INPUT_EVENT", (int)position, 8);
                CurrentItem.Selected = true;
                IndexChangedEvent();
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.ToString());
            }
        }

        public override void GoDown()
        {
            try
            {
                CurrentItem.Selected = false;
                index++;
                if (index >= Items.Count)
                    index = 0;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_INPUT_EVENT", (int)position, 9);
                CurrentItem.Selected = true;
                IndexChangedEvent();
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.ToString());
            }
        }

        public override void Select()
        {
            SelectItem();
        }

        public override void GoBack()
        {
            Focused = false;
        }

        public override void MouseScroll(int dir)
        {
            CurrentItem.Selected = false;
            index += dir;
            if (index < 0)
                index = Items.Count - 1;
            if (index >= Items.Count)
                index = 0;
            CurrentItem.Selected = true;
            IndexChangedEvent();
        }

        public StoreItem CurrentItem => (StoreItem)Items[CurrentSelection];
        public int CurrentSelection
        {
            get => index;
            set
            {
                CurrentItem.Selected = false;
                index = value;
                if (index < 0)
                    index = Items.Count - 1;
                else if (index >= Items.Count)
                    index = 0;
                CurrentItem.Selected = true;
                if (visible)
                    Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)position, index, true, true);
                IndexChangedEvent();
            }
        }

        public void RemoveItem(int id)
        {
            Items.RemoveAt(id);
            if (visible)
            {

            }
        }

        public void SortMissions(Comparison<StoreItem> compare)
        {
            try
            {
                CurrentItem.Selected = false;
                _unfilteredItems = Items.ToList();
                _unfilteredSelection = CurrentSelection;
                Clear();
                List<StoreItem> list = _unfilteredItems.Cast<StoreItem>().ToList();
                list.Sort(compare);
                Items = list.Cast<PauseMenuItem>().ToList();
                if (visible)
                {
                    Populate();
                    ShowColumn();
                }
            }
            catch (Exception ex)
            {
                ResetFilter();
                Debug.WriteLine("ScaleformUI - " + ex.ToString());
            }
        }

        public void FilterMissions(Func<StoreItem, bool> predicate)
        {
            if (predicate == null)
                throw new ArgumentNullException(nameof(predicate));
            try
            {
                _unfilteredItems = Items.ToList();
                _unfilteredSelection = CurrentSelection;
                //_unfilteredTopEdge = topEdge;

                var filteredItems = Items.Cast<StoreItem>().Where(predicate).ToList();

                if (!filteredItems.Any())
                {
                    Debug.WriteLine("^1ScaleformUI - No items were found, resetting the filter");
                    _unfilteredItems.Clear();
                    _unfilteredSelection = 0;
                    return;
                }

                Items[CurrentSelection].Selected = false;
                Clear();

                Items = filteredItems.Cast<PauseMenuItem>().ToList();
                CurrentSelection = 0;

                if (visible)
                {
                    Populate();
                    ShowColumn();
                }
            }
            catch (Exception ex)
            {
                ResetFilter();
                Debug.WriteLine($"^1ScaleformUI - Error filtering menu items: {ex}");
                throw;
            }
        }

        public void ResetFilter()
        {
            try
            {
                if (_unfilteredItems != null && _unfilteredItems.Count > 0)
                {
                    CurrentItem.Selected = false;
                    Clear();
                    Items = _unfilteredItems.ToList();
                    CurrentSelection = _unfilteredSelection;
                    if (visible)
                    {
                        Populate();
                        ShowColumn();
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine("ScaleformUI - " + ex.ToString());
            }
        }

        //private void RefreshColumn()
        //{
        //    if (Parent is MainView lobby)
        //        lobby._pause._lobby.CallFunction("CLEAR_STORE_COLUMN");
        //    else if (Parent is TabView pause && ParentTab.Visible)
        //        pause._pause._pause.CallFunction("CLEAR_PLAYERS_TAB_STORE_COLUMN");
        //    if (Items.Count > 0)
        //    {
        //        isBuilding = true;
        //        int max = Pagination.ItemsPerPage;
        //        if (Items.Count < max)
        //            max = Items.Count;

        //        Pagination.MinItem = Pagination.CurrentPageStartIndex;
        //        if (Pagination.scrollType == ScrollingType.CLASSIC && Pagination.TotalPages > 1)
        //        {
        //            int missingItems = Pagination.GetMissingItems();
        //            if (missingItems > 0)
        //            {
        //                Pagination.ScaleformIndex = Pagination.GetPageIndexFromMenuIndex(Pagination.CurrentPageEndIndex) + missingItems;
        //                Pagination.MinItem = Pagination.CurrentPageStartIndex - missingItems;
        //            }
        //        }
        //        Pagination.MaxItem = Pagination.CurrentPageEndIndex;

        //        for (int i = 0; i < max; i++)
        //        {
        //            if (!Parent.Visible) return;
        //            _itemCreation(Pagination.CurrentPage, i, false, true);
        //        }
        //        Pagination.ScaleformIndex = Pagination.GetScaleformIndex(CurrentSelection);
        //        if (Parent is MainView _lobby)
        //        {
        //            _lobby._pause._lobby.CallFunction("SET_STORE_SELECTION", Pagination.ScaleformIndex);
        //            _lobby._pause._lobby.CallFunction("SET_STORE_QTTY", CurrentSelection + 1, Items.Count);
        //        }
        //        else if (Parent is TabView _pause && ParentTab.Visible)
        //        {
        //            _pause._pause._pause.CallFunction("SET_PLAYERS_TAB_STORE_SELECTION", Pagination.ScaleformIndex);
        //            _pause._pause._pause.CallFunction("SET_PLAYERS_TAB_STORE_QTTY", CurrentSelection + 1, Items.Count);
        //        }
        //        isBuilding = false;
        //    }
        //}


        public void SelectItem()
        {
            StoreItemActivated?.Invoke(CurrentItem, CurrentSelection);
        }
        public void IndexChangedEvent()
        {
            OnIndexChanged?.Invoke(CurrentSelection);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/Elements/Columns/StoreListColumn.lua
StoreListColumn = {}
StoreListColumn.__index = StoreListColumn
setmetatable(StoreListColumn, { __index = PM_Column })
StoreListColumn.__call = function() return "StoreListColumn" end

---@class StoreListColumn
---@field private _label string
---@field private _color SColor
---@field private _isBuilding boolean
---@field private _currentSelection number
---@field private _unfilteredItems table
---@field private _rightLabel string
---@field public Order number
---@field public Parent function
---@field public ParentTab number
---@field public Items table<number, StoreItem>
---@field public OnIndexChanged fun(index: number)
---@field public AddStoreItemItem fun(self: StoreListColumn, item: StoreItem)

function StoreListColumn.New(label)
    local base = PM_Column.New(-1)
    base.Label = label
    base.type = PLT_COLUMNS.MISSION
    base.VisibleItems = 4
    base._unfilteredItems = {} --[[@type table<number, StoreItem>]]
    base._unfilteredSelection = 1
    base.OnIndexChanged = function(index)
    end
    base.OnStoreItemActivated = function(index)
    end
    return setmetatable(base, StoreListColumn)
end

function StoreListColumn:visible()
    return PM_Column.visible(self)
end

function StoreListColumn:SetVisibleItems(maxItems)
    self.VisibleItems = maxItems
    if self:visible() then
        self:Populate()
        self:ShowColumn()
    end
end

function StoreListColumn:ShowColumn()
    if not self:visible() then return end
    PM_Column.ShowColumn(self)
    self:InitColumnScroll(#self.Items >= self.VisibleItems, 1, ScrollType.UP_DOWN, ScrollArrowsPosition.RIGHT)
    self:SetColumnScroll(self.index, #self.Items, self.VisibleItems, "", #self.Items < self.VisibleItems)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_FOCUS", self.position, self.Focused, false, false)
end

function StoreListColumn:Populate()
    if not self:visible() then return end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_DATA_SLOT_EMPTY", self.position)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_MAX_ITEMS", self.position, self.VisibleItems)
    for i=1, #self.Items, 1 do
        self:SetDataSlot(i)
    end
end

---Sets or gets the current selection.
---@param value? number
---@return number | nil
function StoreListColumn:CurrentSelection(value)
    if value == nil then
        return self.index
    else
        self:CurrentItem():Selected(false)
        self.index = value
        if self.index < 1 then
            self.index = #self.Items
        elseif self.index > #self.Items then
            self.index = 1
        end
        self:CurrentItem():Selected(true)
        if self:visible() and self.Focused then
            ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", self.position, self.index - 1, true, true)
        end
        self.OnIndexChanged(self.index)
    end
end

function StoreListColumn:AddItem(item)
    self:AddStoreItem(item)
end

---Removes an item from the column.
---@param item StoreItem 
function StoreListColumn:RemoveSetting(item)
    if item == nil then
        print("^1[ERROR] SettingsListColumn:RemoveSetting() - item is nil")
        return
    end
    for k,v in pairs(self.Items) do
        if v:Label () == item:Label() then
            self:RemoveSlot(k)
        end
    end
end

function StoreListColumn:RemoveItemAt(index)
    if index >#self.Items or index < 1 then return end
    self:RemoveSlot(index)
end

---Add a new item to the column.
---@param item StoreItem
function StoreListColumn:AddStoreItem(item)
    item.ParentColumn = self
    table.insert(self.Items, item)
    if self:visible() and #self.Items < self.VisibleItems then
        local idx = #self.Items
        self:AddSlot(idx)
        self.Items[idx]:Selected(idx == self.index)
    end
end

function StoreListColumn:SetDataSlot(index)
    if self.index > #self.Items then return end
    if self:visible() then
        self:SendItemToScaleform(index)
    end
end
function StoreListColumn:UpdateSlot(index)
    if self.index > #self.Items then return end
    if self:visible() then
        self:SendItemToScaleform(index, true)
    end
end
function StoreListColumn:AddSlot(index)
    if self.index > #self.Items then return end
    if self:visible() then
        self:SendItemToScaleform(index, false, false, true)
    end
end

function StoreListColumn:AddItemAt(item, index)
    table.insert(self.Items, index, item)
    if not self:visible() then return end
    self:AddSlot(index)
    item:Selected(idx == self.index)
end

function StoreListColumn:SendItemToScaleform(i, update, newItem, isSlot)
    if i > #self.Items then return end
    local item = self.Items[i]
    local str = "SET_DATA_SLOT"
    if update then str = "UPDATE_SLOT" end
    if newItem then str = "SET_DATA_SLOT_SPLICE" end
    if isSlot then str = "ADD_SLOT" end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction(str, self.position, i, 0,0, i, 0, item:Enabled(), item.textureDictionary, item.textureName, item.Description)
    if self.position == 0 and i == self.index then
        if item.Panel ~= nil then
            item.Panel:UpdatePanel()
        end
    end
end

function StoreListColumn:GoUp()
    self.Items[self:CurrentSelection()]:Selected(false)
    self.index = self.index - 1
    if self.index < 1 then
        self.index = #self.Items
    end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_INPUT_EVENT", self.position, 8)
    self.Items[self:CurrentSelection()]:Selected(true)
    self.OnIndexChanged(self:CurrentSelection())
end

function StoreListColumn:GoDown()
    self.Items[self:CurrentSelection()]:Selected(false)
    self.index = self.index - 1
    if self.index < 1 then
        self.index = #self.Items
    end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_INPUT_EVENT", self.position, 9)
    self.Items[self:CurrentSelection()]:Selected(true)
    self.OnIndexChanged(self:CurrentSelection())
end

function StoreListColumn:Clear()
    self:ClearColumn()
end

function StoreListColumn:SortStore(compare)
    if not self:visible() then return end
    self.Items[self:CurrentSelection()]:Selected(false)
    if self._unfilteredItems == nil or #self._unfilteredItems == 0 then
        for i, item in ipairs(self.Items) do
            table.insert(self._unfilteredItems, item)
        end
    end
    self._unfilteredSelection = self:Index()
    self:Clear()
    local list = self._unfilteredItems
    table.sort(list, compare)
    self.Items = list
    if self:visible() then
        self:Populate()
        self:ShowColumn()
    end
end

function StoreListColumn:FilterStore(predicate)
    self.Items[self:CurrentSelection()]:Selected(false)
    if self._unfilteredItems == nil or #self._unfilteredItems == 0 then
        for i, item in ipairs(self.Items) do
            table.insert(self._unfilteredItems, item)
        end
    end
    self._unfilteredSelection = self:Index()
    self:Clear()
    local filteredItems = {}
    for i, item in ipairs(self._unfilteredItems) do
        if predicate(item) then
            table.insert(filteredItems, item)
        end
    end
    self.Items = filteredItems
    if self:visible() then
        self:Populate()
        self:ShowColumn()
    end
end

function StoreListColumn:ResetFilter()
    if self._unfilteredItems ~= nil and #self._unfilteredItems > 0 then
        self.Items[self:CurrentSelection()]:Selected(false)
        self:Clear()
        self.Items = self._unfilteredItems
        self:Index(self._unfilteredSelection)
        self._unfilteredItems = {}
        self._unfilteredSelection = 1
        if self:visible() then
            self:Populate()
            self:ShowColumn()
        end
    end
end